README

 LCTuple
 -------

 Marlin package that creates a ROOT TTree with a column wise ntuple
 from LCIO collections.

 F.Gaede, DESY, 2011
 B.Vormwald, DESY, 2012 
 F. Mueller, C. Duerig, A. Ebrahimi, DESY, 2015

-----------------------------------------------------



  Build LCTuple
  -------------

    . /path/to/ilcsoft/installation/v01-XX/init_ilcsoft.sh
    mkdir build
    cd build
    cmake -C $ILCSOFT/ILCSoft.cmake ..
    make install # this should create library: lib/libLCTuple.so
    cd ..


 
 Running LCTuple
 ---------------

    export MARLIN_DLL=$PWD/lib/libLCTuple.so
    cd example
    Marlin --global.LCIOInputFiles=./DST01-06_ppr004_6f_bbqqqq_500_ILD_00_0.001.slcio --global.MaxRecordNumber=1000 lctuple.xml
    

    // make plots with TTree:Draw(), e.g.
    root -l 
    .x /example_plots.C("./lctuple.root")
    
    // see also other examples	
    draw_simhits.C
       - hit maps from sim hits r-z and r-phi

    draw_mcp_vtx.C
       - draw vertices of secondary MCParticles

    draw_vtx_z.C
       - draw tue and reconstructed z-vertex position

    draw_etot.C
       - draw true and reconstructed total energy
 
-------------------------------------------------------------------------------------------

  --------
 | v01-04 |
  --------

   - Added JetBranches which holds branches created from a jet collection
     created by Satoru Jet Finder processor.
     JetBranches has two steerable parameters which are optionally
     set in the steering file:
      * writeExtraParameters: if set to TRUE, the fill function
        calculates some extra parameters using parameters provided
        by the jet finding algorithm. Default value is FALSE.
        See class data members for a list of extra parameters.
      * writeTaggingParameters: if set to TRUE, the fill function
        creates branches for jet parameters coming form flavor tagging
        processor. If no flavor tagging processr is used, this should
        be set to FALSE to prevent errors. Default value is FALSE.
        See class data members for a list of extra parameters.
   - Added isolated lepton branches
   - Added MC particle from relation branches
   - Added MC particle remove overlay branches

  --------
 | v01-03-01 |
  --------

   - add an example for creating an lctuple w/ simhits (F.Gaede)


  --------
 | v01-03 |
  --------
   -  added support for TrackerHits

  --------
 | v01-02 |
  --------
   - (J.List) added some variables: MCParticle daughters: mcda0-mcda4 , Event parameters: evene, evpoe, evpop
             RecoParticles: cluster size: rcncl

  --------
 | v01-01 |
  --------
   -  only change : create the TTree with the name of the processor
                    (used to be hard coded LCTuple)


  --------
 | v01-00 |
  --------

       - first version creates CWNT for 
        MCParticles, ReconstructedParticles, Tracks, Clusters, LCRelations, Vertices

      - optionally also SimTrackerHits and SimCalorimeterHits can
        be written to the ntuple
  


doc/ReleaseNotes.md
# v01-14

* 2022-06-29 Kollassery Swathi Sasikumar ([PR#10](https://github.com/iLCSoft/LCTuple/pull/10))
  - Changed the names of parameter in the event header to the names that are used in MC2020

* 2022-06-28 Thomas Madlener ([PR#13](https://github.com/iLCSoft/LCTuple/pull/13))
  - Make doxygen cmake config work with newer cmake versions (>= 3.17)

# v01-13

* 2021-08-24 Andre Sailer ([PR#12](https://github.com/iLCSoft/LCTuple/pull/12))
  - CI: build against LCG_99python2 gcc8 and LCG_100 gcc10, clang11

* 2021-08-02 Thomas Madlener ([PR#11](https://github.com/iLCSoft/LCTuple/pull/11))
  - Migrate CI to github actions and remove travis CI

* 2020-03-05 gianelle ([PR#9](https://github.com/iLCSoft/LCTuple/pull/9))
  - TrackerHitBranches: add branch to store the time of the tracker hits

# v01-12

* 2019-03-26 Ete Remi ([PR#7](https://github.com/iLCSoft/LCTuple/pull/7))
  - Jet branches: added individual particles collection indices to a new output branch `pfoori`
  - Fixed various warning, making this package warning free

* 2018-05-04 Frank Gaede ([PR#5](https://github.com/iLCSoft/LCTuple/pull/5))
  - add possibility to store arbitrary PID parameters for the ReconstructedParticle branch
      - parameter and branches need to be defined in  `PIDBranchDefinition` parameter of LCTuple
      - see [../example/lctupletof.xml](../example/lctupletof.xml) for details

# v01-11

* 2018-03-13 Marko Petric ([PR#4](https://github.com/iLCSoft/LCTuple/pull/4))
  -  Fix for iLCSoft/LCIO#35

# v01-10

* 2017-11-06 Frank Gaede ([PR#3](https://github.com/ilcsoft/LCTuple/pull/3))
  - add CalorimeterHit::getTime() as 'catim'

# v01-09

* 2017-07-26 Frank Gaede ([PR#2](https://github.com/iLCSoft/LCTuple/pull/2))
  - add collection parameters to output collection in MergeCollections
         - add parameter CollectionParameterIndex
          - copy the collection parameters from the given input collection
             to the output collection
          - note: this is done in addition to the collection parameters from all
             collections that are added under a new name w/ the collection name
             prepended

# v01-08

# v01-08

 S.Sasikumar:
  - add parameter rcnrp ( ReconstructedParticle::getParticles().size() ) to RecoParticle branch
  

# v01-07

# v01-04
* Added JetBranches which holds branches created from a jet collection created by Satoru Jet Finder processor. JetBranches has two steerable parameters which are optionally set in the steering file:
  * writeExtraParameters: if set to TRUE, the fill function calculates some extra parameters using parameters provided by the jet finding algorithm. Default value is FALSE. See class data members for a list of extra parameters.
  * writeTaggingParameters: if set to TRUE, the fill function creates branches for jet parameters coming form flavor tagging processor. If no flavor tagging processr is used, this should be set to FALSE to prevent errors. Default value is FALSE. See class data members for a list of extra parameters.
* Added isolated lepton branches
* Added MC particle from relation branches
* Added MC particle remove overlay branches

# v01-03-01
* add an example for creating an lctuple w/ simhits (F.Gaede)

# v01-03
* added support for TrackerHits

# v01-02
* (J.List) added some variables: MCParticle daughters: mcda0-mcda4 , Event parameters: evene, evpoe, evpop RecoParticles: cluster size: rcncl

# v01-01
* only change : create the TTree with the name of the processor (used to be hard coded LCTuple)

# v01-00
* first version creates CWNT for  MCParticles, ReconstructedParticles, Tracks, Clusters, LCRelations, Vertices
* optionally also SimTrackerHits and SimCalorimeterHits can be written to the ntuple
 


include/JetBranches.h
/*
 * =====================================================================================
 *
 *       Filename:  JetBranches.h
 *
 *    Description:  JetBranches holds branches created from a jet collection
 *					created by Satoru Jet Finder processor.
 *					JetBranches has two steerable parameters which are optionally
 *					set in the steering file:
 *					- writeExtraParameters: if set to TRUE, the fill function
 *					calculates some extra parameters using parameters provided
 *					by the jet finding algorithm. Default value is FALSE.
 *					See class data members for a list of extra parameters.
 *					- writeTaggingParameters: if set to TRUE, the fill function
 *					creates branches for jet parameters coming form flavor tagging
 *					processor. If no flavor tagging processr is used, this should
 *					be set to FALSE to prevent errors. Default value is FALSE.
 *					See class data members for a list of extra parameters.
 *
 *        Version:  1.0
 *        Created:  05/26/2015 02:35:22 PM
 *       Revision:  none
 *       Compiler:  gcc
 *
 *         Author:  Claude Duerig, Felix Mueller, Aliakbar Ebrahimi 
 *   Organization:  DESY
 *
 * =====================================================================================
 */

#ifndef JetBranches_h
#define JetBranches_h 1

/* #####   HEADER FILE INCLUDES   ################################################### */
#include <vector>
#include "LCTupleConf.h" 
#include "UTIL/PIDHandler.h"

#include "CollectionBranches.h"


class TTree ;

namespace EVENT{
  class LCCollection ;
  class LCCEvent ;
}

/** JetBranches holds branches created from a ReconstructedParticle collection
 * 
 * @author F. Gaede, DESY
 * @version $Id: JetBranches.h 4433 2013-01-24 13:09:57Z boehmej $
 */


/*
 * =====================================================================================
 *        Class:  JetBranches
 *  Description:  holds branches created by a SatoruJetFinder processor
 * =====================================================================================
 */
class JetBranches : public CollectionBranches
{
   public:
	  /* ====================  LIFECYCLE     ======================================= */
      JetBranches () {};                        /* constructor */
      JetBranches(const JetBranches&) = delete ;
      JetBranches& operator=(const JetBranches&) = delete ;
	  virtual ~JetBranches() {} ;               /* Destructor */

	  /* ====================  ACCESSORS     ======================================= */

	  // Following two functions are used to access parameters set in a steering file
	  void writeExtraParameters(bool setextraparameters){ _writeExtraParameters = setextraparameters; };
	  void writeTaggingParameters(bool settaggingparameters){ _writeTaggingParameters = settaggingparameters; };

	  /* ====================  MUTATORS      ======================================= */

	  /* ====================  OPERATORS     ======================================= */

	  virtual void initBranches( TTree* tree, const std::string& prefix="" ) ; //const char*  prefix=0) ;
	  virtual void fill(const EVENT::LCCollection* col, EVENT::LCEvent* evt ) ;

   private:
	  /* ====================  METHODS       ======================================= */

	  /* ====================  DATA MEMBERS  ======================================= */

	  bool _writeExtraParameters {} ;               /* Whether to write calculated parameters */
      bool _writeTaggingParameters {} ;             /* Whether to write parameters from tagging processor */

      unsigned int    _nj {} ;                     /* Number of Jets */

	  // Deafault jet parameters
	  float  _jmox[ LCT_JET_MAX ] {} ;             /* Jet X-Momentum */
	  float  _jmoy[ LCT_JET_MAX ] {} ;             /* Jet Y-Momentum */
	  float  _jmoz[ LCT_JET_MAX ] {};             /* Jet Z-Momentum */
	  float  _jmas[ LCT_JET_MAX ] {};             /* Jet Mass */
	  float  _jene[ LCT_JET_MAX ] {};             /* Jet Energy */
	  float  _jcha[ LCT_JET_MAX ] {};             /* Jet Charge */
      float  _jcov0[ LCT_JET_MAX ] {};            /* Jet Covariance matrix element 0 */
      float  _jcov1[ LCT_JET_MAX ] {};            /* Jet Covariance matrix element 1 */
      float  _jcov2[ LCT_JET_MAX ] {};            /* Jet Covariance matrix element 2 */
      float  _jcov3[ LCT_JET_MAX ] {};            /* Jet Covariance matrix element 3 */
      float  _jcov4[ LCT_JET_MAX ] {};            /* Jet Covariance matrix element 4 */
      float  _jcov5[ LCT_JET_MAX ] {};            /* Jet Covariance matrix element 5 */
      float  _jcov6[ LCT_JET_MAX ] {};            /* Jet Covariance matrix element 6 */
      float  _jcov7[ LCT_JET_MAX ] {};            /* Jet Covariance matrix element 7 */
      float  _jcov8[ LCT_JET_MAX ] {};            /* Jet Covariance matrix element 8 */
      float  _jcov9[ LCT_JET_MAX ] {};            /* Jet Covariance matrix element 9 */

	  // Used for tagging parameters
      float  _btag[ LCT_JET_MAX ] {};             /* ? */
      float  _ctag[ LCT_JET_MAX ] {};             /* ? */
      float  _bctag[ LCT_JET_MAX ] {};            /* ? */
      float  _bcat[ LCT_JET_MAX ] {};             /* ? */
      float  _otag[ LCT_JET_MAX ] {};             /* ? */
	  int algo{};                                 /* ? */
	  int ibtag{};                                /* ? */
	  int ictag{};                                /* ? */
	  int iotag{};                                /* ? */
	  int ibcat{};                                /* ? */
	  lcio::PIDHandler *pid{};
      int _jnpid   {};                            /* ? */

	  // Used for extra parameters
	  float  _jevis {};			                /* Jet Visible Energy */
	  float  _jPxvis{};                           /* Jet Momentum (x-component) */
	  float  _jPyvis{};                           /* Jet Momentum (y-component) */
	  float  _jPzvis{};                           /* Jet Momentum (z-component) */
      float  _jPtvis{};                           /* Visible transverse momentum */
	  float  _jmom[ LCT_JET_MAX ]{} ;             /* Jet Total Momentum */
	  float  _jmmax{} ;                           /* Maximum mass of the Jets */
	  float  _jcost[ LCT_JET_MAX ] {};            /* ? */
	  float  _jTheta{};                           /* ? */
      float  _jcosTheta{};                        /* Angle between z-axis and jet */
      float  _jmvis{};                            /* Visible mass */
      float  _jEmiss{};                           /* Missing energy */
      float  _jMmiss{};                           /* Missing Mass */
      float  _jMmissq{};                          /* Missing mass squared */

      unsigned int _njetpfo[ LCT_JET_MAX ] {};             /* Number of PFOs in a jet */
      int _jetpfoori[ LCT_JET_MAX ][ LCT_JET_PARTICLES_MAX ] {} ; /* PFOs collection indexes for each jet */


}; /* -----  end of class JetBranches  ----- */

#endif
